### PR TITLE
Add `Map` type and utility functions to Prelude

### PR DESCRIPTION
**Pull request by Gabriel439**
... as suggested in https://github.com/dhall-lang/dhall-lang/issues/557#issuecomment-495670317

### CODE REVIEW

File: Prelude/Map/keys [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+-}
+
+let Map = ../Map.dhall
+
+let List/map = ../List/map
+
+let keys : ∀(k : Type) → ∀(v : Type) → Map k v → List k
+      = λ(k : Type)
+      → λ(v : Type)
+      → List/map { mapKey : k, mapValue : v } k
```
Comment (philandstuff) [inline Prelude/Map/keys]: Given the frequency of this record type, is it worth capturing it as something like `Map/entry k v`? Then, for example, we could define `Map k v` as `List (Map/entry k v)`

File: Prelude/Map.dhall [hunk 2]
```diff
@@ -0,0 +1,24 @@
+{- This is the canonical way to encode a dynamic list of key-value pairs.
+
+   Tools (such as `dhall-to-json`/`dhall-to-yaml` will recognize values of this
+   type and convert them to maps/dictionaries/hashes in the target language
+
+   For example, `dhall-to-json` converts a Dhall value like this:
+
+   ```
+   [ { mapValue = "foo", mapValue = 1 }
+   , { mapValue = "bar", mapValue = 2 }
```
Comment (Profpatsch) [inline Prelude/Map.dhall]: `mapKey`

File: Prelude/Map/keys (hunk 1 above)
Comment (Profpatsch) [reply Prelude/Map/keys]: `Map/Item`?

Comment (Gabriel439) [reply Prelude/Map/keys]: So I googled for "association list" and the elements of the list are commonly referred to as "pairs", so maybe `Map/Pair`?

Normally the term `Pair` is a bit generic and suggests an arbitrary 2-tuple, but if it's qualified by `Map/` then I think it's probably okay.

Comment (Nadrieril) [reply Prelude/Map/keys]: I feel that `Map/Pair` could still be unclear; `Map/Entry` conveys nicely what it's meant to be used for. Also `Entry` is the terminology used for this usecase [in the Rust ecosystem](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html)

**Comment by singpolyma** [comment]
> I feel that `Map/Pair` could still be unclear; `Map/Entry` conveys nicely what it's meant to be used for.

I agree with Map/Entry

File: Prelude/Map/keys (hunk 1 above)
Comment (jneira) [reply Prelude/Map/keys]: [In java](https://docs.oracle.com/javase/7/docs/api/java/util/Map.Entry.html) it is used for the key/value pair too. +1 for that

**Pull request merged**